From 5bc34aec899eef88a67e6c1e6a6ba4584d4438b1 Mon Sep 17 00:00:00 2001
From: mayi <3363523343@qq.com>
Date: Wed, 21 Oct 2026 18:34:14 +0800
Subject: [PATCH] feat(osdrv/atbm6461): add CVITEK SDIO platform callbacks

Use cvi_sdio_rescan() and existing Linux SDIO enumeration instead of private MMC-class probing. Keep the module powered, leave power_ctrl NULL and do not claim unrelated wake/power GPIOs.
---
 .../realtek/atbm6461/hal_apollo/apollo_plat.h |  2 ++
 .../realtek/atbm6461/hal_apollo/apollo_sdio.c |  4 ++++
 .../atbm6461/hal_apollo/atbm_platform.c       | 19 +++++++++++++++++--
 3 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_plat.h b/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_plat.h
index 7fc66f0e355df13a1892e2ca219e523c602c72fc..9fdd94ff077d33f8762e1c708c31f9c1dd99d4b4 100644
--- a/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_plat.h
+++ b/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_plat.h
@@ -37,6 +37,8 @@
 #define PLATFORM_XM530						(28)
 #define PLATFORM_GK7202V330         		(29)
 
+#define PLATFORM_CVITEK                 (30)
+
 #ifndef  ATBM_WIFI_PLATFORM
 #define ATBM_WIFI_PLATFORM			PLATFORM_XUNWEI
 #endif
diff --git a/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_sdio.c b/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_sdio.c
index ee0aa94c1b82500fe2ebe05b8eeda0c04bb67c9b..88b9bc1d9c7dbf94208242a2cc033e4d0dec75e4 100644
--- a/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_sdio.c
+++ b/extdrv/wireless/realtek/atbm6461/hal_apollo/apollo_sdio.c
@@ -923,6 +923,9 @@ static int atbm_sdio_irq_unsubscribe(struct sbus_priv *self)
 	&& (ATBM_WIFI_PLATFORM != PLATFORM_AMLOGIC_905))
 static int atbm_detect_card(const struct atbm_platform_data *pdata)
 {
+#if (ATBM_WIFI_PLATFORM == PLATFORM_CVITEK)
+	return 0;
+#else
 	/* HACK!!!
 	 * Rely on mmc->class_dev.class set in mmc_alloc_host
 	 * Tricky part: a new mmc hook is being (temporary) created
@@ -990,6 +993,7 @@ platform_dev_err:
 	if(sdio_platform_dev)
 		platform_device_put(sdio_platform_dev);
 	return status;
+#endif
 }
 #endif //PLATFORM_AMLOGIC_S805
 
diff --git a/extdrv/wireless/realtek/atbm6461/hal_apollo/atbm_platform.c b/extdrv/wireless/realtek/atbm6461/hal_apollo/atbm_platform.c
index 71bf3f291708c3596a8792838229cd96328c6874..91ce07df07cb2dd7c3d19b8cde6881726f4332c8 100644
--- a/extdrv/wireless/realtek/atbm6461/hal_apollo/atbm_platform.c
+++ b/extdrv/wireless/realtek/atbm6461/hal_apollo/atbm_platform.c
@@ -17,6 +17,11 @@
 #include "sbus.h"
 #include "apollo_plat.h"
 
+#if (ATBM_WIFI_PLATFORM == PLATFORM_CVITEK)
+#define PLATFORMINF "cv181x-bf01"
+extern int cvi_sdio_rescan(void);
+#endif
+
 #if (ATBM_WIFI_PLATFORM == PLATFORM_XUNWEI)
 #define PLATFORMINF	"xunwei"
 #define CARD_INSERT_GPIO EXYNOS4_GPX3(2)
@@ -224,6 +229,10 @@ static int atbm_platform_insert_crtl(const struct atbm_platform_data *pdata,bool
 {
 	int ret = 0;
 #ifdef SDIO_BUS
+	#if (ATBM_WIFI_PLATFORM == PLATFORM_CVITEK)
+	if (enabled)
+		ret = cvi_sdio_rescan();
+	#endif
 
 	#if (ATBM_WIFI_PLATFORM == PLATFORM_INGENIC)
 	{
@@ -443,13 +452,19 @@ void atbm_plat_free_gpio_irq(const struct atbm_platform_data *pdata,struct sbus_
 #endif  //CONFIG_ATBM_APOLLO_USE_GPIO_IRQ
 #endif
 struct atbm_platform_data platform_data = {
-#if ((ATBM_WIFI_PLATFORM == 22) || (ATBM_WIFI_PLATFORM == 10))
+#if (ATBM_WIFI_PLATFORM == PLATFORM_CVITEK)
+	.mmc_id = NULL,
+#elif ((ATBM_WIFI_PLATFORM == 22) || (ATBM_WIFI_PLATFORM == 10))
 	.mmc_id       = "mmc1",
 #else
 	.mmc_id       = "mmc2",
 #endif
 	.clk_ctrl     = NULL,
-	.power_ctrl   = atbm_power_ctrl,
+#if (ATBM_WIFI_PLATFORM == PLATFORM_CVITEK)
+		.power_ctrl   = NULL,
+#else
+		.power_ctrl   = atbm_power_ctrl,
+#endif
 	.insert_ctrl  = atbm_insert_crtl,
 #if(ATBM_WIFI_PLATFORM == PLATFORM_XUNWEI)
 	.irq_gpio	= EXYNOS4_GPX2(4),
